sycl/doc/SYCL_compiler_and_runtime_design.md

```diff
@@ -42,6 +42,91 @@ back-end. Today middle-end transformations include just a couple of passes:
 - **Back-end** - produces native "device" code in ahead-of-time compilation
 mode.
 
+### SYCL support in clang frontend
+
+SYCL support in clang frontend can be split into the following components:
+
+- Device code outlining. Since SYCL is a single source programming model
+compiler should be able to separate device code from host code.
+
+- Lowering of lambda function objects and named function objects ("SYCL kernel
+functions"). To execute "SYCL kernel functions" on OpenCL devices some
+transformations are required.
+
+- Device code diagnostics.
+
+- Integration header generation.
+
+#### Device code outlining
+
+Here is a code example of a SYCL program that demonstrates compiler outlining
+work:
+
+```C++
+int foo(int x) { return ++x; }
+int bar(int x) { throw std::exception("CPU code only!"); }
+...
+using namespace cl::sycl;
+queue Q;
+buffer<int, 1> a(range<1>{1024});
+Q.submit([&](handler& cgh) {
+      auto A = a.get_access<access::mode::write>(cgh);
+      cgh.parallel_for<init_a>(range<1>{1024}, [=](id<1> index) {
+        A[index] = index[0] * 2 + index[1] + foo(42);
+      });
+    }
+...
+```
+
+SYCL compiler needs to compile lambda exression passed to
```

Comment:
The SYCL compiler needs to compile lambda expressions passed to

Reply:
I understand the paragraph however, perhaps the following is a little clearer:

"In this example, the SYCL compiler needs to compile the lambda expression passed to the `cl::sycl::handler::parallel_for` method, as well as the function foo called from the lambda expression for the device. The compiler must also ignore the bar function when we compile the "device" part of the single source code, as it's unused inside the device portion of the source code."

Hopefully, this is reasonable sounding and doesn't change the intent of the paragraph?

Reply:
Yep, It's sounding better but I'm not sure about expression in brackets:
> device portion of the source code (the contents of the cl::sycl::handler::parallel_for lambda expression)

It makes me feel that **device code only in lambda expression** passed to the `cl::sycl::handler::parallel_for` but device code also in any function called from lambda expression so I rewrote it as:
> portion of the source code (the contents of the lambda expression passed to the
`cl::sycl::handler::parallel_for` and any function called from this lambda
expression).

Reply:
Yes, that sounds better than my original rewording.